src/lib/config.ts

```diff
@@ -0,0 +1,417 @@
+
+    this.graphqlHost = config.strOpt("graphqlHost");
+    this.graphqlPort = config.numberOpt("graphqlPort");
+  }
+
+  makeEndpoint(
+    scope?: string,
+    role?: string
+  ): {
+    secret: string;
+    url: string | undefined;
+    graphqlHost: string | undefined;
+    graphqlPort: number | undefined;
+  } {
+    const secret =
+      this.secret + (scope ? `:${scope}` : "") + ":" + (role ?? "admin");
+
+    return {
+      secret,
+      url: this.url,
+      graphqlHost: this.graphqlHost,
+      graphqlPort: this.graphqlPort,
+    };
+  }
+}
+
+class Stack {
+  /**
+   * The endpoint name to use as a base.
+   */
+  endpoint: string;
+  /**
+   * The database path to use.
+   */
+  database: string | undefined;
+
+  constructor(config: Config) {
+    this.endpoint = config.str("endpoint");
+    this.database = config.strOpt("database");
+  }
+
+  updateDatabase(database: string | undefined) {
+    if (database !== undefined) {
+      this.database = database;
+    }
+  }
+
+  makeSecret(endpoint?: Endpoint, scope?: string, role?: string) {
+    const scope0 = scope ?? this.database;
+    const secret =
+      endpoint?.secret + (scope0 ? `:${scope0}` : "") + ":" + (role ?? "admin");
+
+    return secret;
+  }
+}
+
+const getURLFromConfig = (config: Config): string | undefined => {
+  const url = config.strOpt("url");
+  const scheme = config.strOpt("scheme");
+  const domain = config.strOpt("domain");
+  const port = config.numberOpt("port");
+
+  if (
+    url === undefined &&
+    (domain !== undefined || port !== undefined || scheme !== undefined)
+  ) {
+    const scheme0 = scheme ?? "https";
+    const domain0 = domain ?? "db.fauna.com";
+    const port0 = port ? `:${port}` : "";
+    return `${scheme0}://${domain0}${port0}`;
+  } else {
+    return url;
+  }
+};
+
+/**
+ * Builds the options provided to the faunajs client.
+ * Tries to load the ~/.fauna-shell file and read the default endpoint from there.
+ *
+ * Assumes that if the file exists, it would have been created by fauna-shell,
+ * therefore it would have a defined endpoint.
+ *
+ * Flags like --host, --port, etc., provided by the CLI take precedence over what's
+ * stored in ~/.fauna-shell.
+ *
+ * The --endpoint flag overries the default endpoint from fauna-shell.
+ *
+ * If ~/.fauna-shell doesn't exist, tries to build the connection options from the
+ * flags passed to the script.
+ *
+ * It always expect a secret key to be set in ~/.fauna-shell or provided via CLI
+ * arguments.
+ */
+export const lookupEndpoint = (flags: any, scope: string, role: string) => {
+  const rootConfig = ini.parse(readFileOpt(getRootConfigFile()));
+  const localConfigPath = getLocalConfigFile();
+  const localConfig = localConfigPath
+    ? ini.parse(fs.readFileSync(localConfigPath, "utf8"))
+    : undefined;
+
+  return lookupEndpoint0({
+    flags,
+    rootConfig,
+    localConfig,
+    scope,
+    role,
+  });
+};
+
+export const lookupEndpoint0 = (opts: {
```

Comment:
this is starting to feel like this is doing a lot more than just looking up and endpoint, which I suppose is the end result of what we are doing.  but we're also parsing and validating config.  I'm wondering if it would make sense to have something like a overarching Configuration class, that you create, and it parses and validates all your configuration.  Then on it there is a method to get the endpoint, and that, with everything already parsed, processed and validated no just does the things needed to return the endpoint.

this could take in all the ini config as well as the set of flags and build the correct config object.